core/asset/service_test.go

```diff
@@ -430,9 +467,12 @@ func TestService_GetAssetByID(t *testing.T) {
 			}
 
 			svc := asset.NewService(mockAssetRepo, mocks.NewDiscoveryRepository(t), mocks.NewLineageRepository(t))
-			_, err := svc.GetAssetByID(ctx, tc.ID)
+			actual, err := svc.GetAssetByID(ctx, tc.ID)
+			if tc.Expected != nil {
+				assert.Equal(t, *tc.Expected, actual)
+			}
 			if tc.ExpectedErr != nil {
-				assert.EqualError(t, err, tc.ExpectedErr.Error())
+				assert.ErrorAs(t, err, &tc.ExpectedErr)
```

Comment:
@sudo-suhas  The error that is returned from the mocked repository might not be directly returned as is.
Example

```
{
	Description: `with URN, should return error from repository`,
	ID:          urn,
	Setup: func(ctx context.Context, ar *mocks.AssetRepository) {
		ar.EXPECT().GetByURN(ctx, urn).Return(asset.Asset{}, errors.New("the world exploded"))
	},
	ExpectedErr: errors.New("the world exploded"),
},
```

That test will only pass, if we return error from mock as is. While in implementation, it could've been wrapped.

Reply:
The problem is that `errors.As` only checks that the type is as expected.

```go
// ErrorAs asserts that at least one of the errors in err's chain matches target, and if so, sets target to that error value.
// This is a wrapper for errors.As.
```

This wouldn't verify other aspects like a specific property being set.

Reply:
in that case, should asset by error message (string) then? that should give more confidence right?